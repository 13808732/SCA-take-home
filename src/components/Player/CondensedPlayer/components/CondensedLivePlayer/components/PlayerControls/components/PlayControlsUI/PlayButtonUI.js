import LoadingIcon from 'components/Icons/loading.svg';
import PlayIcon from 'components/Icons/rounded-play.svg';
import { bool, func } from 'prop-types';
import React from 'react';
import styled, { keyframes } from 'styled-components';
import screen from 'styles/helpers/media';
import StopIcon from './StopIcon';

const SpinKeyFrame = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

const AnimatedLoadingIcon = styled(LoadingIcon)`
  animation: ${SpinKeyFrame} 0.75s linear infinite;
`;

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  position: relative;
  cursor: pointer;

  & svg {
    width: 110%;
    height: 110%;
    position: absolute;
  }
`;

const PlayStatus = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  width: 28px;
  height: 28px;
  background-color: ${props => props.theme.primary};
  outline: none;
  appearance: none;
  border: transparent;
  cursor: pointer;

  & svg {
    width: 85%;
    height: 85%;
  }

  ${screen.md} {
    width: 36px;
    height: 36px;
    & svg {
      width: 90%;
      height: 90%;
    }
  }
`;

function PlayButton({ isPlaying, isLoading, onClick }) {
  const getTrackStatusIcon = () => {
    if (isLoading) {
      return <AnimatedLoadingIcon />;
    }

    if (isPlaying) {
      return <StopIcon />;
    }

    return <PlayIcon />;
  };

  const getTitle = () => {
    if (isLoading) {
      return 'Loading';
    }

    if (isPlaying) {
      return 'Stop';
    }

    return 'Play';
  };

  return (
    <Wrapper onClick={onClick}>
      <PlayStatus title={getTitle()}>{getTrackStatusIcon()}</PlayStatus>
    </Wrapper>
  );
}

PlayButton.propTypes = {
  onClick: func,
  isPlaying: bool,
  isLoading: bool,
};

PlayButton.defaultProps = {
  onClick: null,
  isPlaying: false,
  isLoading: false,
};

export default PlayButton;
