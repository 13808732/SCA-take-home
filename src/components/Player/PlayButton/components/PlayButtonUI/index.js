import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import PlayIcon from 'components/Icons/rounded-play.svg';
import PauseIcon from 'components/Icons/pause.svg';
import StopIcon from 'components/Icons/stop.svg';
import LoadingIcon from 'components/Icons/loading.svg';

const StyledButton = styled.button`
  transition: transform 0.2s ease, background-color 0.2s ease;
  background-color: transparent;
  border: 0;
  padding: 0;
  cursor: pointer;
  border-radius: 50%;
  height: ${props => props.size};

  & svg {
    width: ${props => props.size};
    height: ${props => props.size};
  }

  &:hover,
  &:active {
    outline: none;
  }

  &:focus {
    outline: none;
  }
`;

const SpinKeyFrame = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

const AnimatedLoadingIcon = styled(LoadingIcon)`
  animation: ${SpinKeyFrame} 0.75s linear infinite;
`;

function PlayButtonUI({ isPlaying, isStreaming, isLoading, onClick, variant, ...props }) {
  const getButtonSize = () => {
    if (variant === 'xxs') {
      return '22px';
    }
    if (variant === 'xs') {
      return '33px';
    }
    if (variant === 's') {
      return '36px';
    }

    if (variant === 'm') {
      return '40px';
    }

    if (variant === 'l') {
      return '55px';
    }

    return '70px';
  };

  const getTrackStatusIcon = () => {
    if (isLoading) {
      return <AnimatedLoadingIcon />;
    }
    // stop icon for streaming
    if (!isPlaying && isStreaming) {
      return <StopIcon />;
    }

    // pause icon for non-streaming
    if (isPlaying && !isStreaming) {
      return <PauseIcon />;
    }

    return <PlayIcon />;
  };

  const getTitle = () => {
    if (isLoading) {
      return 'Loading';
    }

    if (!isPlaying && isStreaming) {
      return 'Stop';
    }

    if (isPlaying && !isStreaming) {
      return 'Pause';
    }

    return 'Play';
  };

  return (
    <StyledButton size={getButtonSize()} onClick={onClick} title={getTitle()} {...props}>
      {getTrackStatusIcon()}
    </StyledButton>
  );
}

PlayButtonUI.propTypes = {
  onClick: PropTypes.func,
  isStreaming: PropTypes.bool,
  isPlaying: PropTypes.bool,
  isLoading: PropTypes.bool,
  variant: PropTypes.oneOf(['xxs', 'xs', 's', 'm', 'l']),
};

PlayButtonUI.defaultProps = {
  variant: 'l',
  onClick: null,
  isStreaming: false,
  isPlaying: false,
  isLoading: false,
};

export default PlayButtonUI;
