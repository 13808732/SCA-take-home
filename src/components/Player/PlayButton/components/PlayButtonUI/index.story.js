import React from 'react';
import { storiesOf } from '@storybook/react';
import PlayButtonUI from './index';

storiesOf('Play button', module)
  .addWithJSX('l, paused', () => (
    <PlayButtonUI
      className=""
      isPlaying={false}
      isLoading={false}
      variant="l"
      onClick={() => {}}
    />
  ))
  .addWithJSX('m, playing', () => (
    <PlayButtonUI
      className=""
      isPlaying
      isLoading={false}
      variant="m"
      onClick={() => {}}
    />
  ))
  .addWithJSX('s, loading', () => (
    <PlayButtonUI
      className=""
      isPlaying={false}
      isLoading
      variant="s"
      onClick={() => {}}
    />
  ))
  .addWithJSX('xs, playing', () => (
    <PlayButtonUI
      className=""
      isPlaying
      isLoading={false}
      variant="xs"
      onClick={() => {}}
    />
  ))
  .addWithJSX('xl, loading', () => (
    <PlayButtonUI
      className=""
      isPlaying={false}
      isLoading
      variant="xl"
      onClick={() => {}}
    />
  ));
