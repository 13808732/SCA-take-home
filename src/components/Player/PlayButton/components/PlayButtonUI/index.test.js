import React from 'react';
import renderer from 'react-test-renderer';
import PlayButtonUI from './index';

describe('<PlayButtonUI /> component', () => {
  const props = {
    onClick: jest.fn(),
    isPlaying: false,
    isLoading: false,
    variant: 'xs',
  };

  test('it should render a PauseIcon component', () => {
    const component = renderer.create(<PlayButtonUI {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it should render a StopIcon component', () => {
    props.catchupMiniPlayer = false;

    const component = renderer.create(<PlayButtonUI {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
