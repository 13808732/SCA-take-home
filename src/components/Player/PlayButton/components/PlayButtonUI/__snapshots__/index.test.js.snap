// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayButtonUI /> component it should render a PauseIcon component 1`] = `
<button
  className="sc-bdnylx cMhJU"
  onClick={[MockFunction]}
  size="33px"
  title="Play"
>
  <svg
    height="70"
    viewBox="0 0 70 70"
    width="70"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M49.422 36.789L25.894 48.553A2 2 0 0 1 23 46.763V23.237a2 2 0 0 1 2.894-1.789l23.528 11.764a2 2 0 0 1 0 3.578z"
      fill="#FFF"
      fillRule="evenodd"
    />
  </svg>
</button>
`;

exports[`<PlayButtonUI /> component it should render a StopIcon component 1`] = `
<button
  className="sc-bdnylx cMhJU"
  onClick={[MockFunction]}
  size="33px"
  title="Play"
>
  <svg
    height="70"
    viewBox="0 0 70 70"
    width="70"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M49.422 36.789L25.894 48.553A2 2 0 0 1 23 46.763V23.237a2 2 0 0 1 2.894-1.789l23.528 11.764a2 2 0 0 1 0 3.578z"
      fill="#FFF"
      fillRule="evenodd"
    />
  </svg>
</button>
`;
